// @kbn/config/target/index.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPluginSearchPaths = exports.LegacyObjectToConfigAdapter = exports.Env = exports.ObjectToConfigAdapter = exports.hasConfigPathIntersection = exports.isConfigPath = exports.ConfigService = exports.getConfigFromFiles = exports.RawConfigService = exports.configDeprecationFactory = exports.applyDeprecations = void 0;
var deprecation_1 = require("./deprecation");
Object.defineProperty(exports, "applyDeprecations", { enumerable: true, get: function () { return deprecation_1.applyDeprecations; } });
Object.defineProperty(exports, "configDeprecationFactory", { enumerable: true, get: function () { return deprecation_1.configDeprecationFactory; } });
var raw_1 = require("./raw");
Object.defineProperty(exports, "RawConfigService", { enumerable: true, get: function () { return raw_1.RawConfigService; } });
Object.defineProperty(exports, "getConfigFromFiles", { enumerable: true, get: function () { return raw_1.getConfigFromFiles; } });
var config_service_1 = require("./config_service");
Object.defineProperty(exports, "ConfigService", { enumerable: true, get: function () { return config_service_1.ConfigService; } });
var config_1 = require("./config");
Object.defineProperty(exports, "isConfigPath", { enumerable: true, get: function () { return config_1.isConfigPath; } });
Object.defineProperty(exports, "hasConfigPathIntersection", { enumerable: true, get: function () { return config_1.hasConfigPathIntersection; } });
var object_to_config_adapter_1 = require("./object_to_config_adapter");
Object.defineProperty(exports, "ObjectToConfigAdapter", { enumerable: true, get: function () { return object_to_config_adapter_1.ObjectToConfigAdapter; } });
var env_1 = require("./env");
Object.defineProperty(exports, "Env", { enumerable: true, get: function () { return env_1.Env; } });
var legacy_1 = require("./legacy");
Object.defineProperty(exports, "LegacyObjectToConfigAdapter", { enumerable: true, get: function () { return legacy_1.LegacyObjectToConfigAdapter; } });
var plugins_1 = require("./plugins");
Object.defineProperty(exports, "getPluginSearchPaths", { enumerable: true, get: function () { return plugins_1.getPluginSearchPaths; } });
