// @kbn/utils/target/streams/index.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFilterStream = exports.createReplaceStream = exports.createMapStream = exports.createConcatStream = exports.createPromiseFromStreams = exports.createReduceStream = exports.createListStream = exports.createSplitStream = exports.createIntersperseStream = exports.concatStreamProviders = void 0;
var concat_stream_providers_1 = require("./concat_stream_providers");
Object.defineProperty(exports, "concatStreamProviders", { enumerable: true, get: function () { return concat_stream_providers_1.concatStreamProviders; } });
var intersperse_stream_1 = require("./intersperse_stream");
Object.defineProperty(exports, "createIntersperseStream", { enumerable: true, get: function () { return intersperse_stream_1.createIntersperseStream; } });
var split_stream_1 = require("./split_stream");
Object.defineProperty(exports, "createSplitStream", { enumerable: true, get: function () { return split_stream_1.createSplitStream; } });
var list_stream_1 = require("./list_stream");
Object.defineProperty(exports, "createListStream", { enumerable: true, get: function () { return list_stream_1.createListStream; } });
var reduce_stream_1 = require("./reduce_stream");
Object.defineProperty(exports, "createReduceStream", { enumerable: true, get: function () { return reduce_stream_1.createReduceStream; } });
var promise_from_streams_1 = require("./promise_from_streams");
Object.defineProperty(exports, "createPromiseFromStreams", { enumerable: true, get: function () { return promise_from_streams_1.createPromiseFromStreams; } });
var concat_stream_1 = require("./concat_stream");
Object.defineProperty(exports, "createConcatStream", { enumerable: true, get: function () { return concat_stream_1.createConcatStream; } });
var map_stream_1 = require("./map_stream");
Object.defineProperty(exports, "createMapStream", { enumerable: true, get: function () { return map_stream_1.createMapStream; } });
var replace_stream_1 = require("./replace_stream");
Object.defineProperty(exports, "createReplaceStream", { enumerable: true, get: function () { return replace_stream_1.createReplaceStream; } });
var filter_stream_1 = require("./filter_stream");
Object.defineProperty(exports, "createFilterStream", { enumerable: true, get: function () { return filter_stream_1.createFilterStream; } });
